### PR TITLE
Convert [ParticleProto] to [Plan.Particle].

### PR DESCRIPTION
**Pull request by bgogul**
One more to go convert recipe.proto to the Kotlin classes!

### CODE REVIEW

File: java/arcs/core/data/proto/ParticleProtoDecoder.kt [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    }
+    return handleProto.decodeAsHandleConnection(handleMode)
+}
+
+/** Converts a [ParticleProto] into a [Particle]. */
+fun ParticleProto.decode(context: DecodingContext): Particle {
+    val particleSpec = context.particleSpecs[specName]
+    require(particleSpec != null) {
+        "ParticleSpec '$specName' not found in decoding context."
+    }
```
Comment (jasonwyatt) [inline java/arcs/core/data/proto/ParticleProtoDecoder.kt]:
```suggestion
    val particleSpec = requireNotNull(context.particleSpecs[specName]) {
        "ParticleSpec '$specName' not found in decoding context."
    }
```

File: java/arcs/core/data/proto/ParticleProtoDecoder.kt [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    context: DecodingContext
+): HandleConnection {
+    val handleSpec = particleSpec.connections[name]
+    require(handleSpec != null) {
+        "HandleConnectionSpec '$name' not found in ParticleSpec '${particleSpec.name}'."
+    }
+    val handleProto = context.handleProtos[handle]
+    require(handleProto != null) {
+        "HandleProto for '$handle' not found when decoding ParticleProto '${particleSpec.name}'."
+    }
```
Comment (jasonwyatt) [inline java/arcs/core/data/proto/ParticleProtoDecoder.kt]:
```suggestion
    val handleSpec = requireNotNull(particleSpec.connections[name]) {
        "HandleConnectionSpec '$name' not found in ParticleSpec '${particleSpec.name}'."
    }
    val handleProto = requireNotNull(context.handleProtos[handle]) {
        "HandleProto for '$handle' not found when decoding ParticleProto '${particleSpec.name}'."
    }
```

File: java/arcs/core/data/proto/ParticleProtoDecoder.kt [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+/** Converts a [ParticleProto] into a [Particle]. */
+fun ParticleProto.decode(context: DecodingContext): Particle {
+    val particleSpec = context.particleSpecs[specName]
+    require(particleSpec != null) {
+        "ParticleSpec '$specName' not found in decoding context."
+    }
+    val handleConnections = connectionsList.map {
+        it.handle to it.decode(particleSpec, context)
+    }.toMap()
```
Comment (jasonwyatt) [inline java/arcs/core/data/proto/ParticleProtoDecoder.kt]: Consider using `associateBy` or one of the related methods.

File: javatests/arcs/core/data/proto/ParticleProtoDecoderTest.kt [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        .setStorageKey(ramdiskStorageKey)
+        .build()
+    val readConnectionSpec = HandleConnectionSpec("data", Direction.READS, TypeVariable("data"))
+    val readerSpec = ParticleSpec("Reader", mapOf("data" to readConnectionSpec), "ReaderLocation")
+    val writeConnectionSpec = HandleConnectionSpec("data", Direction.WRITES, TypeVariable("data"))
+    val writerSpec = ParticleSpec("Writer", mapOf("data" to writeConnectionSpec), "WriterLocation")
+    val context = DecodingContext(
+        particleSpecs = mapOf("Reader" to readerSpec, "Writer" to writerSpec),
+        handleProtos = mapOf("thing" to thingHandleProto)
+    )
```
Comment (jasonwyatt) [inline javatests/arcs/core/data/proto/ParticleProtoDecoderTest.kt]: These can go at the top of the file (they're not in a companion object, so they're just instance fields)

**Comment by bgogul** [comment]
> Some readability suggestions, but nothing huge :)

Thanks, Jason. Addressed your comments. I added some more tests involving multiple connections. PTAL.

File: java/arcs/core/data/proto/ParticleProtoDecoder.kt (hunk 3 above)
Comment (bgogul) [reply java/arcs/core/data/proto/ParticleProtoDecoder.kt]: Unfortunately, the decoded `HandleConnection` does not have the `name` that I can use to `associateBy`. I have marked adding `name` as one of the TODOs.

**Comment by bgogul** [comment]
Thanks for the review, folks! However, I need to abandon this PR. The `Plan.Particle` does not allow me the capture information that is  necessary for type inference and other analysis. recipe.proto has all the inforamation. Therefore, I will introduce a `Recipe` type in Kotlin for the purposes of the analysis.  Once types are resolved, the recipe can be converted to a `Plan`.

**Pull request closed without merging**